test-utils/src/main/java/com/criteo/publisher/view/WebViewLookup.java

```diff
@@ -55,19 +54,28 @@
    * Look inside the given view for {@link android.webkit.WebView}. Then if exactly one webview is
    * found, then its HTML content is fetch.
    * <p>
-   * If no webview is found, then the future will fail with a {@link java.util.NoSuchElementException}.
-   * And if more than one webview are found, then the future will fail with a {@link
-   * TooManyWebViewsException}.
+   * If no webview is found, then the future will sleep a little and try again. And if more than one
+   * webview are found, then the future will fail with a {@link TooManyWebViewsException}.
    *
    * @param root the view to look into
    * @return either the HTML content of the found webview, either an error state.
    */
   public Future<String> lookForHtmlContent(View root) {
     CompletableFuture<String> future = new CompletableFuture<>();
+    lookForHtmlContent(root, future);
+    return future;
+  }
+
+  private void lookForHtmlContent(View root, CompletableFuture<String> future) {
     List<WebView> webViews = lookForWebViews(root);
 
     if (webViews.isEmpty()) {
-      future.completeExceptionally(new NoSuchElementException());
+      // Sleep a little and retry to search for webview
+      executor.submit(() -> {
+        Thread.sleep(20);
+        lookForHtmlContent(root, future);
```

Comment:
Isn't there a risk of endless loop there?

Reply:
Yes but there is no risk of stack overflow, as a new stack is given to the new thread. And this method is wrapped in a future which can be cancelled, making the sleep throw an InterruptedException.
Moreover, almost all tests are using the MockDependencyProviderRule which setup a thread killing the test if not fulfilled after 1 minute.